File: solutions/python/solution_02_pairwise.py
ALPHABET = 26

def removeAnagramsAndSubAnagrams(words: list[str]) -> list[str]:
    if not words:
        return []

    groups = {}
    for word in words:
        key = tuple(freq_vector(word))
        groups.setdefault(key, []).append(word)

    unique_groups = {k: v for k, v in groups.items() if len(v) == 1}

    if not unique_groups:
        return []

    vecs = list(unique_groups.keys())

    maximal = [candidate for candidate in vecs
               if not any(dominates(other, candidate) for other in vecs)]

    return [unique_groups[vec][0] for vec in maximal]


def freq_vector(word: str) -> list[int]:
    f = [0] * ALPHABET
    for c in word:
        if 'a' <= c <= 'z':
            f[ord(c) - ord('a')] += 1
    return f


def dominates(a: tuple[int, ...], b: tuple[int, ...]) -> bool:
    if a is b:
        return False
    if a == b:
        return False

    dominated = True
    strict = False

    for i in range(ALPHABET):
        if a[i] < b[i]:
            dominated = False
            break
        if a[i] > b[i]:
            strict = True

    return dominated and strict

File: solutions/kotlin/Solution02_Pairwise.kt
private const val ALPHABET_02 = 26

fun removeAnagramsAndSubAnagrams02(words: List<String>): List<String> {
    if (words.isEmpty()) return emptyList()

    val groups = words.groupBy { freqVector02(it).toList() }
    val uniqueGroups = groups.filter { it.value.size == 1 }

    if (uniqueGroups.isEmpty()) return emptyList()

    val vecs = uniqueGroups.keys.map { it.toIntArray() }
    val maximal = vecs.filter { candidate ->
        vecs.none { other -> dominates02(other, candidate) }
    }

    return maximal.map { uniqueGroups[it.toList()]!!.first() }
}

private fun freqVector02(word: String): IntArray {
    val freq = IntArray(ALPHABET_02)
    for (c in word) {
        if (c in 'a'..'z') freq[c - 'a']++
    }
    return freq
}

private fun dominates02(a: IntArray, b: IntArray): Boolean {
    if (a === b) return false
    if (a.contentEquals(b)) return false

    var strict = false
    for (i in 0 until ALPHABET_02) {
        if (a[i] < b[i]) return false
        if (a[i] > b[i]) strict = true
    }
    return strict
}
